benchmark/initializer.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require "bundler/inline"
+
+gemfile do
+  source "https://rubygems.org"
+  gem "dry-initializer", "~> 3.1"
+  gem "dry-types", "~> 1.5"
+  gem "awesome_print"
+  gem "benchmark-ips"
+  gem "pry"
+  gem "ruby-prof"
+  gem "sorbet"
+  gem "sorbet-runtime"
+  gem "definition", path: File.expand_path("../.", __dir__)
+end
+
+class SorbetUseCase
+  extend T::Sig
+
+  sig { params(phone: String, admin: T::Boolean).void }
+  def initialize(phone:, admin: false)
+    self.phone = phone
+    self.admin = admin
+  end
+
+  private
+
+  attr_accessor :phone, :admin
+end
+
+class DryUseCase
+  extend Dry::Initializer
+
+  option :phone, Dry::Types["strict.string"]
+  option :admin, Dry::Types["strict.bool"], default: false, optional: true
+end
+
+class DefinitionUseCase
+  include Definition::Initializer
+
+  required :phone, Definition.Type(String)
+  optional :admin, Definition.Boolean, default: false
+end
+
+puts "Benchmark with valid input data:"
+valid_data = { phone: "+49 3424 234234" }
+
+Benchmark.ips do |x|
+  x.config(time: 5, warmup: 2)
+
+  x.report("sorbet") do
+    SorbetUseCase.new(**valid_data)
+  end
+
+  x.report("definition") do
```

Comment:
unfortunately definition looses here, as dry-initializer actually compiles the initializer code instead of running through its full fledged validation code. However its still in the 0.00something ms range on my machine

Reply:
Benchmark with valid input data:
Warming up --------------------------------------
              sorbet    82.329k i/100ms
          definition    37.577k i/100ms
          dry-struct    88.061k i/100ms
Calculating -------------------------------------
              sorbet    852.495k (± 3.7%) i/s -      4.281M in   5.029531s
          definition    374.358k (± 3.0%) i/s -      1.879M in   5.023693s
          dry-struct    858.138k (± 2.4%) i/s -      4.315M in   5.031203s

Comparison:
          dry-struct:   858137.7 i/s
              sorbet:   852494.7 i/s - same-ish: difference falls within error
          definition:   374357.5 i/s - 2.29x  (± 0.00) slower

Benchmark with invalid input data:
Warming up --------------------------------------
              sorbet     8.068k i/100ms
          definition     4.753k i/100ms
          dry-struct    10.624k i/100ms
Calculating -------------------------------------
              sorbet     77.992k (± 6.1%) i/s -    395.332k in   5.087320s
          definition     47.174k (± 5.5%) i/s -    237.650k in   5.051701s
          dry-struct     96.592k (± 7.2%) i/s -    488.704k in   5.085971s

Comparison:
          dry-struct:    96591.8 i/s
              sorbet:    77992.0 i/s - 1.24x  (± 0.00) slower
          definition:    47173.6 i/s - 2.05x  (± 0.00) slower

Reply:
Changelog needs to be adapted